mllib/src/main/scala/org/apache/spark/ml/FuncTransformer.scala

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml
+
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream, ObjectInputStream, ObjectOutputStream}
+
+import scala.reflect.runtime.universe.{typeOf, TypeTag}
+
+import org.apache.hadoop.fs.Path
+
+import org.apache.spark.annotation.{DeveloperApi, Since}
+import org.apache.spark.ml.FuncTransformer.FuncTransformerWriter
+import org.apache.spark.ml.util._
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.parser.CatalystSqlParser
+import org.apache.spark.sql.types.DataType
+
+/**
+ * :: DeveloperApi ::
+ * A wrapper to allow easily creation of simple data manipulation for DataFrame.
+ * Note that FuncTransformer supports serialization via scala ObjectOutputStream and may not
+ * guarantee save/load compatibility between different scala version.
+ */
+@DeveloperApi
+@Since("2.3.0")
+class FuncTransformer [IN, OUT: TypeTag] @Since("2.3.0") (
+    @Since("2.3.0") override val uid: String,
+    @Since("2.3.0") val func: IN => OUT,
+    @Since("2.3.0") val outputDataType: DataType
+  ) extends UnaryTransformer[IN, OUT, FuncTransformer[IN, OUT]] with DefaultParamsWritable {
+
+  @Since("2.3.0")
+  def this(fx: IN => OUT, outputDataType: DataType) =
+    this(Identifiable.randomUID("FuncTransformer"), fx, outputDataType)
+
+  @Since("2.3.0")
+  def this(fx: IN => OUT) =
+    this(Identifiable.randomUID("FuncTransformer"), fx,
+      CatalystSqlParser.parseDataType(typeOf[OUT].typeSymbol.name.decodedName.toString))
```

Comment:
Thanks for looking into it. I'll try with more test cases but I know this will not work in all the cases.

I'm hesitating about whether we should keep the interface, it's really convenient though. I think maybe adding some check and then throws friendly exception if the output Type cannot be inferred directly. 

I'll send out the unit tests that I already tried.